app/agent/runtime/qemu/qemu.service.test.ts

```diff
@@ -0,0 +1,387 @@
+          },
+          async ({ ssh, runtimePid }) => {
+            const now = Date.now();
+            await waitForPromises([
+              execSshCmd({ ssh }, ["sh", "-c", "sleep 1; echo 'a'"]).catch(() => {
+                sshCommandExitedWithin = Date.now() - now;
+                sshDone(void 0);
+              }),
+              (async () => {
+                await sleep(500);
+                process.kill(runtimePid, "SIGKILL");
+              })(),
+            ]);
+          },
+        ),
+      ).rejects.toThrow();
+      await sshDonePromise;
+      expect(sshCommandExitedWithin).not.toBe(null);
+      expect(sshCommandExitedWithin).toBeLessThan(5000);
+    })(),
+);
+
+// Run those tests only on alpine, this test does not depend on the guest distro but on the guest kernel
+test.concurrent.each(
+  [testCases[0]].flatMap(([name, remoteTag]) => [
+    [1, 128, name, remoteTag],
+    [1, 256, name, remoteTag],
+    [2, 128, name, remoteTag],
+    [2, 256, name, remoteTag],
+  ]),
+)(
+  `Creates vm with %d cores and %d MiB ram on %s`,
+  async (vcpuCount, memSizeMib, _name, remoteTag) =>
+    provideBlockRegistry(async ({ injector, runId, brService }) => {
+      const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+      const osCt = await brService.createContainer(osIm, "osCt");
+      const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
+      const instance = injector.resolve(Token.QemuService)(runId);
+      await instance.withRuntime(
+        {
+          ssh: {
+            username: "root",
+            password: "root",
+          },
+          vcpuCount,
+          memSizeMib,
+          fs: { "/": osB },
+          cleanupAfterStop: true,
+          shouldPoweroff: true,
+        },
+        async ({ ssh }) => {
+          const cpuInfo = (await execSshCmd({ ssh }, ["cat", "/proc/cpuinfo"])).stdout;
+          expect(cpuInfo).toMatch(new RegExp(`^cpu cores.*?${vcpuCount}$`, "gm"));
+          // Ram is more tricky and containers need a different test here
+          // Assert that the memory block size on the guest is 128MB
+          // block_size_bytes is in hex notation
+          await expect(
+            execSshCmd({ ssh }, ["cat", "/sys/devices/system/memory/block_size_bytes"]),
+          ).resolves.toHaveProperty("stdout", "8000000");
+          // Now check the amount of memory blocks on the system
+          expect(
+            (await execSshCmd({ ssh }, ["ls", "/sys/devices/system/memory/"])).stdout
+              .split("\n")
+              .filter((m) => m.startsWith("memory")),
+          ).toHaveLength(memSizeMib / 128);
+        },
+      );
+    })(),
+);
+
+test.concurrent(
+  `Doesn't hang forever if qemu fails to start`,
+  provideBlockRegistry(async ({ injector, runId, brService }) => {
+    const osCt = await brService.createContainer(void 0, "emptyCt", { mkfs: true, sizeInGB: 64 });
+    const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
```

Comment:
hide?

Reply:
It's handled globally: https://github.com/hocus-dev/hocus/pull/93/files#diff-6728b317769e3b07f2f0926e1b27c3dadcd3958b1eb186a33b3cc8b3234f6047R162